src/main/kotlin/io/github/mojira/arisa/ModuleExecutor.kt

```diff
@@ -0,0 +1,334 @@
+import io.github.mojira.arisa.modules.HideImpostorsModule
+import io.github.mojira.arisa.modules.KeepPrivateModule
+import io.github.mojira.arisa.modules.ModuleError
+import io.github.mojira.arisa.modules.ModuleResponse
+import io.github.mojira.arisa.modules.OperationNotNeededModuleResponse
+import io.github.mojira.arisa.modules.PiracyModule
+import io.github.mojira.arisa.modules.RemoveNonStaffMeqsModule
+import io.github.mojira.arisa.modules.RemoveTriagedMeqsModule
+import io.github.mojira.arisa.modules.ReopenAwaitingModule
+import io.github.mojira.arisa.modules.RevokeConfirmationModule
+import net.rcarz.jiraclient.Issue
+import net.rcarz.jiraclient.JiraClient
+import net.sf.json.JSONObject
+import java.text.SimpleDateFormat
+import java.util.Timer
+import kotlin.concurrent.schedule
+
+class ModuleExecutor(
+    private val jiraClient: JiraClient,
+    private val config: Config
+) {
+    private val attachmentModule: AttachmentModule
+    private val chkModule: CHKModule
+    private val crashModule: CrashModule
+    private val emptyModule: EmptyModule
+    private val keepPrivateModule: KeepPrivateModule
+    private val futureVersionModule: FutureVersionModule
+    private val hideImpostorsModule: HideImpostorsModule
+    private val piracyModule: PiracyModule
+    private val removeNonStaffMeqsModule: RemoveNonStaffMeqsModule
+    private val removeTriagedMeqsModule: RemoveTriagedMeqsModule
+    private val reopenAwaitingModule: ReopenAwaitingModule
+    private val revokeConfirmationModule: RevokeConfirmationModule
+
+    private val ticketCache = mutableListOf<String>()
+    private val ticketTimer = Timer("RemoveCachedTicket", true)
+
+    private fun executeModule(
+        moduleConfig: Arisa.Modules.ModuleConfigSpec,
+        queryCache: MutableMap<String, List<Issue>>,
+        processedIssues: MutableMap<String, Boolean>,
+        executeModule: (Issue) -> Pair<String, Either<ModuleError, ModuleResponse>>
+    ) {
+        val projects = config[Arisa.Issues.projects]
+            .filter { config[moduleConfig.whitelist] == null || config[moduleConfig.whitelist]!!.contains(it) }
+            .joinToString(",")
+        val resolutions = config[moduleConfig.resolutions].joinToString(",") { "\"$it\"" }
+        val cachedTickets = if (ticketCache.isEmpty()) "" else "AND key not in (${ticketCache.joinToString(",")})"
+        val combinedJql = "project in ($projects) $cachedTickets AND resolution in ($resolutions) AND (${config[moduleConfig.jql]})"
+
+        val issues = queryCache[combinedJql] ?: jiraClient
+            .searchIssues(combinedJql)
+            .issues
+            .map { jiraClient.getIssue(it.key, "*all", "changelog") } // Get issues again to retrieve all fields
+            .filter { issue ->
+                // Ignore issues where last action was a resolve
+                val latestChange = issue.changeLog.entries.lastOrNull()
+
+                latestChange == null || // There is actually a entry
```

Comment:
This should be != as far as I know

Reply:
If there is no latest change, the filter method needs to return true.
These are exactly the inverse of what used to be in ArisaMain